src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.HttpContentAsyncStream.cs

```diff
@@ -0,0 +1,412 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http
+{
+    // libcurl has a pull-based model and needs to be able to read data on-demand
+    // from the request content stream.  HttpContent provides a ReadAsStreamAsync method
+    // that would theoretically enable this, however it's actually not intended for such
+    // scenarios and has some deficiencies in this regard.  For example, in certain scenarios,
+    // we may need to retry an HTTP request, which means sending the data again.  In such
+    // situations, libcurl will ask us to rewind the stream.  We can do this with ReadAsStreamAsync,
+    // but only if the returned stream supports seeking.  If it doesn't support seeking, we
+    // could ask for a new stream again via calling ReadAsStreamAsync again, but it doesn't
+    // do content consumption detection like you get with HttpContent.CopyToAsync.  CopyToAsync
+    // is the preferred, supported mechanism for a handler getting data from the HttpContent,
+    // and is what WinHttpHandler uses, but it also provides a push-based model.  In order to
+    // be able to pull from a push-based model, we need an adapter stream that straddles
+    // the divide.  This implementation provides that.
+
+    internal partial class CurlHandler : HttpMessageHandler
+    {
+        /// <summary>
+        /// Provides a stream that lets CurlHandler read the data from an
+        /// HttpContent via HttpContent's CopyToAsync.
+        /// </summary>
+        private sealed class HttpContentAsyncStream : Stream
+        {
+            /// <summary>Cached task that contains the Int32 value 0.</int></summary>
+            private static readonly Task<int> s_zeroTask = Task.FromResult(0);
+
+            /// <summary>Object used to synchronize all activity on the stream.</summary>
+            private readonly object _syncObj = new object();
+            /// <summary>The HttpContent to read from.</summary>
```

Comment:
Perhaps useful to mention that HttpClient ensures HttpContent is valid till the HTTP response is received